### DISCUSSION TITLE
Switch from source while streaming

### DISCUSSION THREAD

**Discussion by florianmlr** [pedroSG94/RootEncoder#1051] [Q&A]
Hello Pedro, does this library allow us to switch from source (eg: a file to openglview with camera) while streaming ?

Additionnaly, I try to make this library record from a different quality with opengl, but I still struggle how to add a second encoder, i tried to follow theses steps : https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/issues/967 but i couldn't achieve it, do you have a sample ?

**Comment by pedroSG94** [comment]
Hello,



> does this library allow us to switch from source (eg: a file to openglview with camera) while streaming?

No, it is not developed but it is considered to be in the future (no date for now).



> Additionnaly, I try to make this library record from a different quality with opengl, but I still struggle how to add a second encoder, i tried to follow theses steps : #967 but i couldn't achieve it, do you have a sample ?

I haven't an example for this because I don't want to implement it.
Add a second encoder will reduce the performance because you need to render 3 surfaces (preview, stream, record) instead of 2 (preview and the shared for stream and record), obvious CPU usage for encoding process and camera must be opened in the higher resolution and rescale the image for lower resolution selected.
So make an example is the same to develop a feature that I will drop. I prefer use my time doing fixs or features to add in the project.

**Comment by florianmlr** [comment]
Thank you for your answer.
I will try to implement the switch, do you think that for this feature I will need an encoder for each source (as a video width and height from file source will not be the same as the camera source) ? This will have an impact on performance too I guess.

**Reply by pedroSG94** [reply]
No, you don't need it.

If you use a video with different resolution than camera you will need rescale video source before send it to VideoEncoder.
If you are using OpenGl with FromFile mode, the surface used by the decoder to render video is selected here:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/FromFileBase.java#L377

This means that you maybe need modify OffScreenGlThread and/or OpenGlView to adapt video resolution to the first resolution selected. Anyway, I recommend you first implement change the video source (Basically replace CameraManager to VideoDecoder dinamically). After that, you can start with video resolution modifications.

Also, you need make sure that you audio use the same sample rate in both cases.

**Reply by florianmlr** [reply]
Alright so if i understand it well the idea is to replace dynamically the surface from the openglview (in case of opengl) with the surface from the video encoder ?

**Reply by pedroSG94** [reply]
No,

The idea is replace CameraManager to VideoDecoder dinamically.
If you are using OpenGl CameraManager or VideoDecoder must render surface provider by glInterface. If not, you must render surface provided by VideoEncoder.

**Comment by florianmlr** [comment]
I have an additional question, I have a source that provides YUV i420 data frame, is there a simple way to render it in your openglview ? Do I have to create an other manager (like cameramanager) for this source ?
